### PR TITLE
Upgrade to OpenAPI 3

### PR DESCRIPTION
**Pull request by namoscato**
**⚠️ This is currently a work-in-progress.**

Resolves https://github.com/nelmio/NelmioApiDocBundle/issues/1443

## Details

* Upgrades [zircote/swagger-php](https://github.com/zircote/swagger-php) from `^2.0.12` to `^3.0.2`
* Upgrades [api-platform/core](https://github.com/api-platform/core) from `^2.1.0` to `^2.4.2` (for added OpenAPI 3 support in `2.4.0`)

## Next Steps

- [ ] Update todos in `RouteMetadataDescriber::describe`
- [ ] Update controller action annotations / functional tests (ideally this can be split into more explicit, parallel items)
- [ ] Audit OA `UNDEFINED` checks (i.e. `FormModelDescriber::parseForm` todo, `WebTestCase`)

### CODE REVIEW

File: Tests/Functional/FunctionalTest.php [hunk 1]
```diff
@@ -15,13 +15,13 @@ class FunctionalTest extends WebTestCase
 {
     public function testConfiguredDocumentation()
     {
-        $this->assertEquals('My Default App', $this->getSwaggerDefinition()->info->title);
-        $this->assertEquals('My Test App', $this->getSwaggerDefinition('test')->info->title);
+        $this->assertEquals('My Default App', $this->getOpenApiDefinition()->info->title);
+        $this->assertEquals('My Test App', $this->getOpenApiDefinition('test')->info->title);
     }
```
Comment (namoscato) [inline Tests/Functional/FunctionalTest.php]: My approach to updating functional tests involved starting here, commenting out most `ApiController` methods, and tackling one at a time. Therefore, many `ApiController` (and other controllers) methods currently have invalid annotations.

File: Tests/Functional/Controller/ApiController.php [hunk 2]
```diff
@@ -60,38 +59,33 @@ public function swaggerAction()
...
-     *     @SWG\Schema(
-     *         type="array",
-     *         @SWG\Items(ref=@Model(type=User::class))
-     *     )
+     *     @OA\JsonContent(@OA\Schema(type="array", @OA\Items(ref=@Model(type=User::class))))
      * )
-     * @SWG\Tag(name="implicit")
+     * @OA\Tag(name="implicit")
      */
     public function implicitSwaggerAction()
```
Comment (namoscato) [inline Tests/Functional/Controller/ApiController.php]: This is currently where I am stuck: `AbstractAnnotations` is [logging a warning](https://github.com/zircote/swagger-php/blob/3.0.2/src/Annotations/AbstractAnnotation.php#L226) when `User` properties are merged:

```
/usr/bin/php /Users/nickamoscato/dev/git/NelmioApiDocBundle/vendor/symfony/phpunit-bridge/bin/simple-phpunit --bootstrap /Users/nickamoscato/dev/git/NelmioApiDocBundle/vendor/autoload.php --configuration /Users/nickamoscato/dev/git/NelmioApiDocBundle/phpunit.xml.dist --filter "/(::testConfiguredDocumentation)( .*)?$/" Nelmio\ApiDocBundle\Tests\Functional\FunctionalTest /Users/nickamoscato/dev/git/NelmioApiDocBundle/Tests/Functional/FunctionalTest.php --teamcity --cache-result-file=/Users/nickamoscato/dev/git/NelmioApiDocBundle/.phpunit.result.cache
PHPUnit 7.4.5 by Sebastian Bergmann and contributors.

Testing Nelmio\ApiDocBundle\Tests\Functional\FunctionalTest

Multiple definitions for @OA\Property()->default
     Using: \Nelmio\ApiDocBundle\Nelmio\ApiDocBundle\Tests\Functional\Entity\User->generate() in /Users/nickamoscato/dev/git/NelmioApiDocBundle/ApiDocGenerator.php on line 67
  Skipping: \Nelmio\ApiDocBundle\Tests\Functional\Entity\User->id in /Users/nickamoscato/dev/git/NelmioApiDocBundle/Tests/Functional/Entity/User.php
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/vendor/zircote/swagger-php/src/Logger.php:39
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/vendor/zircote/swagger-php/src/Logger.php:61
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/vendor/zircote/swagger-php/src/Annotations/AbstractAnnotation.php:226
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/ModelDescriber/Annotations/SwgAnnotationsReader.php:85
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/ModelDescriber/Annotations/AnnotationsReader.php:48
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/ModelDescriber/ObjectModelDescriber.php:72
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/Model/ModelRegistry.php:90
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/ApiDocGenerator.php:76
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/Tests/Functional/WebTestCase.php:121
 /Users/nickamoscato/dev/git/NelmioApiDocBundle/Tests/Functional/FunctionalTest.php:18
```

My initial thought was around `zircote/swagger-php`'s `UNDEFINED` constant updates to default annotation property values, but I am not entirely sure how to proceed.

Comment (namoscato) [reply Tests/Functional/Controller/ApiController.php]: This _was_ a missed `UNDEFINED` check in `SwgAnnotationsReader::getPropertyName`, fixed in 1911b627ec6bae49439cde0af5f62b8d72f92a92.

File: Tests/Functional/FunctionalTest.php [hunk 3]
```diff
@@ -31,8 +31,8 @@ public function testFetchArticleAction()
     {
         $operation = $this->getOperation('/api/article/{id}', 'get');
 
-        $response = $this->getResponse($operation, '200');
-        $this->assertEquals('#/definitions/Article', $response->schema->ref);
+        $response = $this->getOperationResponse($operation, '200');
+        $this->assertEquals('#/components/schema/Article', $response->content['application/json']->schema->ref);
```
Comment (namoscato) [inline Tests/Functional/FunctionalTest.php]: I am thinking something like this should be true, but [`ApiController::fetchArticleAction`'s `JsonContent`](https://github.com/namoscato/NelmioApiDocBundle/blob/7b16d00f6f75da2f0e5e60fd158ad3b857c5f571/Tests/Functional/Controller/ApiController.php#L38) does not seem to be getting merged into `Response` as the moment 🤔

**Comment by maxhelias** [comment]
Awesome, it should also manage an alias for compatibility or provide a major release with a update notice :)

**Comment by volbrene** [comment]
After this task is finished is it possible to use multiple server/host Urls in one doc?

**Comment by ChrisB9** [comment]
Is there any news on this and/or do you need assistance finalizing it?

**Comment by namoscato** [comment]
@ChrisB9, any help would be much appreciated! This pull request represents the current state, which is still not fully functional as I have been working through tests.

I am currently stuck on https://github.com/nelmio/NelmioApiDocBundle/pull/1555#discussion_r349889125. I have some more time now that we are into the New Year and can provide some more thoughts around that specific context this week.

**Comment by ChrisB9** [comment]
Okay i'll spin up an environment and then we can see in what way i can help you best

**Comment by misiekR** [comment]
Is there any news on this?

**Comment by galvani** [comment]
@namoscato I can help too. I would love to see it working.

File: composer.json [hunk 4]
```diff
@@ -40,11 +40,12 @@
...
         "doctrine/common": "^2.4",
 
-        "api-platform/core": "^2.1.0",
+        "api-platform/core": "^2.4.7",
         "friendsofsymfony/rest-bundle": "^2.0",
         "willdurand/hateoas-bundle": "^1.0",
         "jms/serializer": "^1.13",
-        "jms/serializer-bundle": "^2.0"
+        "jms/serializer-bundle": "^2.0",
+        "symfony/serializer": "^4.3"
```
Comment (namoscato) [inline composer.json]: I didn't spend too much time looking into this yet, but the normalizers in the current state of [api-platform/core](https://github.com/api-platform/core) (`DateTimeIdentifierDenormalizer`, `IntegerDenormalizer`) seem to be incompatible with the 5.0 `DenormalizerInterface` type-hint additions (https://github.com/symfony/serializer/commit/bfc494904936664389a6efd4ee278b1231c7059e), so I just required an older version for now.

File: Tests/Functional/FunctionalTest.php (hunk 3 above)
Comment (namoscato) [reply Tests/Functional/FunctionalTest.php]: This is still an issue.

Interestingly, replacing `@OA\JsonContent` with `@OA\MediaType` in this controller action seems to get included correctly, but I'm not sure why `@OA\JsonContent` is not getting merged. I would expect its [nested context](https://github.com/zircote/swagger-php/blob/3.0.3/src/Processors/MergeJsonContent.php#L25) to be its parent `@OA\Response`, but it is `null`.

Perhaps this is a side effect of the nested `@Model`? Tangentially related: I'm wondering if `@Model`'s properties should be defaulted to `OpenApi\Annotations\UNDEFINED`?

**Comment by namoscato** [comment]
Closing in favor of https://github.com/nelmio/NelmioApiDocBundle/pull/1623. Thank you again, @FilipBenco!

**Pull request closed without merging**